File: ios/flutter_blue_background/Sources/flutter_blue_background/FbbLog.swift
import Foundation
import os.log

enum FbbLogLevel: Int, CaseIterable {
  case none = 0
  case error = 1
  case warning = 2
  case info = 3
  case debug = 4
  case verbose = 5
}

enum FbbLog {
  private static let subsystem = "com.sparkleo.flutter_blue_background"
  private static let log = OSLog(subsystem: subsystem, category: "FBB")

  private static var level: FbbLogLevel = .debug
  private static var bannerShown = false

  static func setLevel(_ ordinal: Int) {
    level = FbbLogLevel(rawValue: ordinal) ?? .debug
    if level != .none {
      printBanner()
    }
  }

  static func error(_ message: String) { log(.error, message) }
  static func warning(_ message: String) { log(.warning, message) }
  static func info(_ message: String) { log(.info, message) }
  static func debug(_ message: String) { log(.debug, message) }
  static func verbose(_ message: String) { log(.verbose, message) }

  private static func printBanner() {
    guard !bannerShown else { return }
    bannerShown = true
    let banner = """
    ══════
      FBB
    ══════
    """
    os_log("%{public}@", log: log, type: .info, banner)
  }

  private static func log(_ messageLevel: FbbLogLevel, _ message: String) {
    guard level != .none else { return }
    guard messageLevel.rawValue <= level.rawValue else { return }
    printBanner()
    let prefix: String
    switch messageLevel {
    case .error: prefix = "✖ ERROR  │"
    case .warning: prefix = "⚠ WARN   │"
    case .info: prefix = "ℹ INFO   │"
    case .debug: prefix = "● DEBUG  │"
    case .verbose: prefix = "› VERBOSE│"
    case .none: return
    }
    let line = "\(prefix) \(message)"
    let osType: OSLogType = switch messageLevel {
    case .error: .error
    case .warning: .fault
    case .info: .info
    default: .debug
    }
    os_log("%{public}@", log: log, type: osType, line)
  }
}

File: android/src/main/kotlin/com/sparkleo/flutter_blue_background/flutter_blue_background/FbbLog.kt
package com.sparkleo.flutter_blue_background.flutter_blue_background

import android.util.Log

/**
 * Centralized, level-filtered logging for flutter_blue_background on Android.
 *
 * All messages are tagged [TAG] and prefixed with a styled level label so they
 * are easy to filter in logcat (`adb logcat -s FBB`).
 */
enum class FbbLogLevel {
    NONE,
    ERROR,
    WARNING,
    INFO,
    DEBUG,
    VERBOSE,
}

object FbbLog {
    private const val TAG = "FBB"

    @Volatile
    var level: FbbLogLevel = FbbLogLevel.DEBUG
        private set

    @Volatile
    private var bannerShown = false

    fun setLevel(ordinal: Int) {
        level = FbbLogLevel.entries.getOrElse(ordinal) { FbbLogLevel.DEBUG }
        if (level != FbbLogLevel.NONE) {
            printBanner()
        }
    }

    fun error(message: String) = log(FbbLogLevel.ERROR, message)

    fun warning(message: String) = log(FbbLogLevel.WARNING, message)

    fun info(message: String) = log(FbbLogLevel.INFO, message)

    fun debug(message: String) = log(FbbLogLevel.DEBUG, message)

    fun verbose(message: String) = log(FbbLogLevel.VERBOSE, message)

    private fun printBanner() {
        if (bannerShown) return
        bannerShown = true
        Log.i(
            TAG,
            """
            |══════
            |  FBB
            |══════
            """.trimMargin(),
        )
    }

    private fun log(level: FbbLogLevel, message: String) {
        if (this.level == FbbLogLevel.NONE) return
        if (level.ordinal > this.level.ordinal) return
        printBanner()
        val prefix = when (level) {
            FbbLogLevel.ERROR -> "✖ ERROR  │"
            FbbLogLevel.WARNING -> "⚠ WARN   │"
            FbbLogLevel.INFO -> "ℹ INFO   │"
            FbbLogLevel.DEBUG -> "● DEBUG  │"
            FbbLogLevel.VERBOSE -> "› VERBOSE│"
            FbbLogLevel.NONE -> return
        }
        val line = "$prefix $message"
        when (level) {
            FbbLogLevel.ERROR -> Log.e(TAG, line)
            FbbLogLevel.WARNING -> Log.w(TAG, line)
            FbbLogLevel.INFO -> Log.i(TAG, line)
            else -> Log.d(TAG, line)
        }
    }
}
